### PR TITLE
feat: configurable global name 

### LINKED ISSUE #1792: Hiding references to __nuxt in the scripts and the template ID
I've been reading the source code and found out that `window.__nuxt__` and the id for the App component `#__nuxt` are hardcoded. For security reasons I would like to hide what technology is powering my site, do you have plans on having those as options or should I work on my own fork? 

<div align="right"><sub><em>This feature request is available on <a href="https://nuxtjs.cmty.io">Nuxt.js</a> community (<a href="https://nuxtjs.cmty.io/nuxt/nuxt.js/issues/c1605">#c1605</a>)</em></sub></div>

### PR DESCRIPTION
**Pull request by galvez**
Picking up on https://github.com/nuxt/nuxt.js/pull/3522 which's been abandoned for some time.

## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
Permits running multiple Nuxt apps with different names on the same page by introducing a new `globalName` configuration option, along with granular global options (`globals`) based on `globalName` by default.

Resolves #1792

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by pi0** [comment]
I suggest instead of using a single `globalName`, add a configuration to `nuxt.config` to allow customizing different global names. This helps to avoid workarounds like `._on<%= _.capitalize(globalName) %>Loaded` and also the ability to only customize it when needed.

Config example:

```js
globals: {
 rootId: `__nuxt`,
 vuePrototype: `$nuxt`,
 global: `window.$nuxt`,
 state: `window.__NUXT__`,
 readyCbs: `window._nuxtReadyCbs `,
 onLoaded: `window._onNuxtLoaded`,
}
```

In the example above, I've extracted `window.` parts into the configuration too. So we can have even more low-level control over where exactly global callbacks and instance is injected!

File: lib/app/App.js
Comment (pi0) [inline lib/app/App.js]: Why we need to make this internal variable customizable?

File: lib/app/index.js
Comment (pi0) [inline lib/app/index.js]: I suggest to something like `globals.plugin` or `globals.pluginPrefix` instead of directly using `globalName`

File: lib/app/index.js
Comment (pi0) [inline lib/app/index.js]: Instead of `state.state` can't we use something like `context`?

File: lib/common/options.js
Comment (pi0) [inline lib/common/options.js]: What if flatten `globals.callbacks` into `globals.readyCallback`, etc. This way we can eliminate this for loop. (If more root keys like this are added we have to repeat this logic)

File: lib/common/options.js
Comment (pi0) [inline lib/common/options.js]: I suggest moving this into a getter function inside the Builder module. This way, modules have a chance to override options. Also, we don't mutate options anymore.

File: lib/app/index.js
Comment (galvez) [reply lib/app/index.js]: I like this suggestion, will change ASAP.

File: lib/common/options.js
Comment (galvez) [reply lib/common/options.js]: Thanks, I'll work something out.

File: lib/common/options.js
Comment (galvez) [reply lib/common/options.js]: @pi0 I was a little uneasy about it because I feel `callbacks.something` is cleaner than `somethingCallback`, given we've been following an hierarchical style, so it made sense to try and group these things semantically. But as long as it's only a couple of callbacks, I see no harm. Will change before we merge.

**Review by pi0 — APPROVED** [review]
👌 Sexy

**Review by TheAlexLichter — APPROVED** [review]
Really well done :clap:

**Comment by manniL** [comment]
The only thing we might want to consider is that this might break `Nuxt` detection for dev tools, right? :thinking:

**Review by aldarund — APPROVED** [review]

**Review by qm3ster — APPROVED** [review]

**Comment by pi0** [comment]
Waiting for CI

**Comment by axetroy** [comment]
also ``nuxt-progress`` and ``nuxt-link-active`` class name for element

**Comment by monsterboom** [comment]
nice~But the documentation need to be updated~I just set 'globalName' in 'nuxt.config.js', but the 
progress dom cant take effect.

**Comment by manniL** [comment]
@monsterboom Yeah, that's right. (https://github.com/nuxt/docs/issues/842)

Make sure you are on version 2.2!

**Comment by monsterboom** [comment]
@manniL yes,i'm sure, i mean, the dom '<div id="__nuxt"></div>' is changed to my 'globalName', but '<div class="nuxt-progress"></div>' is not changed.  I  npm install lastest just now. I checked package.json and node_modules and the nuxt version was 2.2.0 . Another problem, when i set 'globalName', the browser console a error: 'Cannot read property 'constructor' of undefined' .

**Comment by pi0** [comment]
>  When i set 'globalName', the browser console a error: 'Cannot read property 'constructor' of undefined'

What value do you use for `globalName`?

**Comment by monsterboom** [comment]
the value just my project name "epweb". i found the git tag 2.2.0, the file nuxt-loading.vue is not be changed like in the b1f7123 , https://github.com/nuxt/nuxt.js/blob/v2.2.0/lib/app/components/nuxt-loading.vue

**Comment by sam3d** [comment]
@monsterboom I've found this issue appears to go away when I disable Vue DevTools (as per #4296)

File: lib/app/App.js
Comment (Roriz) [inline lib/app/App.js]: What do you think in apply `pluginPrefix` instead of only `true`?
Will still be a truly and external resources can find the Vue instance (like a vuedevtools).

**Comment by edum18** [comment]
Hello, I have multiple separated apps and need to implement what was discussed here. I've read [docs](https://nuxtjs.org/api/configuration-global-name/) but can't understand how to show two apps on the same page. I'll be very grateful if anyone can write a specific example

**Pull request merged**
